server/src/main/java/io/crate/execution/engine/indexing/GroupRowsByShard.java

```diff
@@ -125,18 +125,30 @@ public void accept(ShardedRequests<TReq, TItem> shardedRequests, Row row) {
 
             String id = rowShardResolver.id();
             TItem item = itemFactory.apply(id);
-            String indexName = indexNameResolver.get();
-            String routing = rowShardResolver.routing();
-            String sourceUri = sourceUriInput.value();
-            Long lineNumber = lineNumberInput.value();
-
-            RowSourceInfo rowSourceInfo = RowSourceInfo.emptyMarkerOrNewInstance(sourceUri, lineNumber);
-            ShardLocation shardLocation = getShardLocation(indexName, id, routing);
-            long sizeEstimate = estimateRowSize.applyAsLong(row);
-            if (shardLocation == null) {
-                shardedRequests.add(item, sizeEstimate, indexName, routing, rowSourceInfo);
-            } else {
-                shardedRequests.add(item, sizeEstimate, shardLocation, rowSourceInfo);
+
+            if (item instanceof ShardUpsertRequest.Item upsertRequestItem
+                &&
+                upsertRequestItem.insertValues() != null
+                &&
+                upsertRequestItem.insertValues().length > 0
+                && upsertRequestItem.insertValues()[0] != null) {
```

Comment:
I think `GroupRowsByShard` shouldn't be aware of concrete `ShardRequest.Item` implementations as it would break encapsulation. 

You could add a new property to the base class that tells consumers if the item succeeded or if it should be skipped.

Reply:
done in https://github.com/crate/crate/pull/12229/commits/6ea66a1780aca2f43bb4ba1f0073b45210436ca2